src/NUnitFramework/framework/Internal/Reflect.cs

```diff
@@ -226,5 +226,89 @@ public static object InvokeMethod( MethodInfo method, object fixture, params obj
+#else
+        /// <summary>
+        /// <para>
+        /// Selects the ultimate shadowing property just like <see langword="dynamic"/> would,
+        /// rather than throwing <see cref="AmbiguousMatchException"/>
+        /// for properties that shadow properties of a different property type
+        /// which is what <see cref="Type.GetProperty(string, BindingFlags)"/> does.
+        /// </para>
+        /// <para>
+        /// If you request both public and nonpublic properties, every public property is preferred
+        /// over every nonpublic property. It would violate the principle of least surprise for a
+        /// derived class’s implementation detail to be chosen over the public API for a type.
+        /// </para>
+        /// </summary>
+        /// <param name="type">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
+        /// <param name="name">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
+        /// <param name="bindingFlags">See <see cref="Type.GetProperty(string, BindingFlags)"/>.</param>
+#endif
+        public static PropertyInfo GetUltimateShadowingProperty(Type type, string name, BindingFlags bindingFlags)
+        {
+            Guard.ArgumentNotNull(type, nameof(type));
+            Guard.ArgumentNotNull(name, nameof(name));
+
+            if ((bindingFlags & BindingFlags.DeclaredOnly) != 0)
+            {
+                // If you're asking us to search a hierarchy but only want properties declared in the given type,
+                // you're in the wrong place but okay:
+                return type.GetProperty(name, bindingFlags);
+            }
+
+            if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == (BindingFlags.Public | BindingFlags.NonPublic))
+            {
+                // If we're searching for both public and nonpublic properties, search for only public first
+                // because chances are if there is a public property, it would be very surprising to detect the private shadowing property.
+
+                for (var publicSearchType = type; publicSearchType != null; publicSearchType = publicSearchType.GetTypeInfo().BaseType)
+                {
+                    var property = publicSearchType.GetProperty(name, (bindingFlags | BindingFlags.DeclaredOnly) & ~BindingFlags.NonPublic);
+                    if (property != null) return property;
+                }
+
+                // There is no public property, so may as well not ask to include them during the second search.
+                bindingFlags &= ~BindingFlags.Public;
+            }
+
+            for (var searchType = type; searchType != null; searchType = searchType.GetTypeInfo().BaseType)
+            {
+                var property = searchType.GetProperty(name, bindingFlags | BindingFlags.DeclaredOnly);
+                if (property != null) return property;
+            }
```

Comment:
AFAICT we can extract the for-loops into a helper method that will accept a Type and BindingFlags and return a property or null, but I'm unsure whether it is worth the work.

Reply:
I guess we could. I'd be more happy about doing that if we could use C# 7 local functions because otherwise e'd have to decide things like whether to apply `| BindingFlags.DeclaredOnly` inside the method (since it's coupled to the particular loop operation) and whether to use guard conditions.

Reply:
Let us just keep it as it is, then we can always improve it later when we get to C# 7 (or newer).